v-next/hardhat/src/internal/builtin-plugins/test/task-action.ts

```diff
@@ -50,9 +51,9 @@ const runAllTests: NewTaskActionFunction<TestActionArguments> = async (
     }
 
     if (subtask.options.has("noCompile")) {
-      await subtask.run({ testFiles: files, noCompile: true });
+      await subtask.run({ testFiles: files, noCompile: true, grep });
     } else {
-      await subtask.run({ testFiles: files });
+      await subtask.run({ testFiles: files, grep });
```

Comment:
To avoid duplication and potential future bug, I'd have outside of the conditional something like `arguments = {testFiles: files, grep}` Then `if(options.has('noCompile')) {arguments['noCompile'] = true} `